Add new section to record Technical Decisions and added deployment pi… (#85)
ministryofjustice/form-builder: technical/deployments-imagineering-session-2.md

# Deployments Imagineering session 2
Created by Tomas Destefi 02 April 2020

> On the last two retros we talked about that our deployment needs a lot of improvements. So 02/04 Me, Brendan and Matt did the session 2 of the Deploy imagineering session.

Reviewer: In the last two retros we talked about how our deployments need a lot of improvements
```suggestion
In the last two retros we talked about how our deployments need a lot of improvements. So 02/04 Me, Brendan and Matt did the session 2 of the Deploy imagineering session.
```

Reviewer 2: 🙈

## Proposal
We want to improve our deployment in all the following applications (sort by level of priority):
1. The submitter app (application responsible to handle submissions)
2. The runner app (application responsible to show the forms to the end user)
3. The user datastore app (application responsible to handle the user data)
4. The user filestore app (application responsible to handle user uploads)
5. The pdf generator app (application responsible to generate pdfs for submitter)
6. The service token cache app (application responsible to make all APIs to communicate with each other)
7. The publisher app (app responsible of publishing the forms)

## Benefits of this work
When this work is finished we:

> 1. We will have more confident on our deployments.

Reviewer: ~confident on~ => confidence in

2. We will deploy apps with more quality.
3. We are adding extra checks before and after it goes live so we can make sure everything worked.
4. Low risk releases.
5. We substantially reduce the cost of making and delivering incremental changes to software by eliminating many of the fixed costs associated with the release process.
6. Faster Mean Time To Resolution.
7. More test reliability.
8. We can implement Continuous Delivery.

## The first part of the work - Submitter
We decided to start implementing the pipeline we envision on the submitter app.

You can see step by step of the deployment pipeline (from left to right with some considerations).

> ![alt text](TD01-deployment.png "Image of the deployment pipeline in context of theis document.")

Reviewer: ~alt text~ => deployment pipeline ?

~theis~ => this

8. [https://trello.com/c/RIBLnCYY/408-watch-pods-being-restarted-in-production-for-the-formbuilder-platform-live-dev-and-formbuilder-platform-live-production](https://trello.com/c/RIBLnCYY/408-watch-pods-being-restarted-in-production-for-the-formbuilder-platform-live-dev-and-formbuilder-platform-live-production)
9. [https://trello.com/c/4hPTxwtd/410-add-healthcheck-step-to-submitter-production-environment](https://trello.com/c/4hPTxwtd/410-add-healthcheck-step-to-submitter-production-environment)
10. [https://trello.com/c/qBo0i1wZ/411-add-smoke-tests-to-test-environment-after-the-deploy-production-is-complete](https://trello.com/c/qBo0i1wZ/411-add-smoke-tests-to-test-environment-after-the-deploy-production-is-complete)
11. [https://trello.com/c/fwgQaChQ/412-alert-channel-about-submitter-pipeline-result](https://trello.com/c/fwgQaChQ/412-alert-channel-about-submitter-pipeline-result)
12. [https://trello.com/c/nFlzcHv7/413-branches-can-deploy-to-test-on-submitter](https://trello.com/c/nFlzcHv7/413-branches-can-deploy-to-test-on-submitter)

## After the work above is done

> After the work above is done we need to apply to other repositories like the Runner, the file store, datastore, and so on.

Reviewer: ```suggestion
After the work above is done we need to apply the same changes to other repositories like the Runner, the file store, datastore, and so on.
```

## Considerations for the developers
1. Don't overengineer but consider the same pipeline process to be applied to other repos like filestore, pdf generator so when doing for submitter the other could be easier to plug in.
2. Branches can deploy to test (and stop there)
3. Consider that we could leave CircleCI in the long term

> 4. Branches can deploy to test (and stop there)

Reviewer: 2 and 4 are the same?

5. In order to give us the confidence we need to test form via PDF/ CSVs / JSON endpoints (and submit each of those) (specific for submitter)

> 6. After this work is started me (Tomas, Brendan and Matt) will review how is going what we agreed.

Reviewer: ~me (Tomas, Brendan and Matt)~=> we  👑 😛
